Привет, меня зовут Люман Аблаев. Сегодня я проверю твой проект.
<br> Дальнейшее общение будет происходить на "ты" если это не вызывает никаких проблем.
<br> Желательно реагировать на каждый мой комментарий ('исправил', 'не понятно как исправить ошибку', ...)
<br> Пожалуйста, не удаляй комментарии ревьюера, так как они повышают качество повторного ревью.

Комментарии будут в <font color='green'>зеленой</font>, <font color='blue'>синей</font> или <font color='red'>красной</font> рамках:

<div class="alert alert-block alert-success">
<b>Успех:</b> Если все сделано отлично
</div>

<div class="alert alert-block alert-info">
<b>Совет: </b> Если можно немного улучшить
</div>

<div class="alert alert-block alert-danger">
<b>Ошибка:</b> Если требуются исправления. Работа не может быть принята с красными комментариями.
</div>


<font color='orange' style='font-size:24px; font-weight:bold'>Общее впечатление</font>
* Спасибо за то, что прислал такую хорошую работу, мне она очень понравилась
- Я оставил некоторые советы, надеюсь они будут полезными или интересными.
* Радует, что код написан довольно чисто, читабельно и аккуратно. Приятно проверять такие работы.
* Тебе удалось успешно справиться с текущей задачей. Поздравляю!
* Но есть маленький недочет - отсутствие итогового вывода и я думаю у тебя не займет много времени дополнить свою работу.

<font color='orange' style='font-size:24px; font-weight:bold'>Общее впечатление[2]</font>
* Спасибо за оперативность.
- Недочет исправлен - работа полностью корректна.
- Удачи в дальнейшем обучении.

# Рекомендация тарифов

В вашем распоряжении данные о поведении клиентов, которые уже перешли на эти тарифы (из проекта курса «Статистический анализ данных»). Нужно построить модель для задачи классификации, которая выберет подходящий тариф. Предобработка данных не понадобится — вы её уже сделали.

Постройте модель с максимально большим значением *accuracy*. Чтобы сдать проект успешно, нужно довести долю правильных ответов по крайней мере до 0.75. Проверьте *accuracy* на тестовой выборке самостоятельно.

<div class="alert alert-block alert-success">
<b>Успех:</b> Спасибо за описание проекта
</div>


## Откройте и изучите файл

<div class="alert alert-block alert-success">
<b>Успех:</b> Отлично, что все импорты собраны в верхней части ноутбука! Если у того, кто будет запускать твой ноутбук будут отсутствовать некоторые библиотеки, то он это увидит сразу, а не в процессе!
  
</div>

In [39]:
import pandas as pd
from pathlib import Path
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

from sklearn.tree import DecisionTreeClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.dummy import DummyClassifier
from sklearn.model_selection import GridSearchCV

import seaborn as sbn 

df = pd.read_csv('/datasets/users_behavior.csv')
df.head(5)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
calls       3214 non-null float64
minutes     3214 non-null float64
messages    3214 non-null float64
mb_used     3214 non-null float64
is_ultra    3214 non-null int64
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


<div class="alert alert-info">
    
<b>Совет</b>
Также важно изучить корреляции в данных. Так мы можем оценить взаимосвязи в данных и влияние параметров на целевой признак. Рекомендую посмотреть метод `sns.pairplot` с параметром hue, в который мы передаем целевой признак. Будет здорово, если ты сможешь это реализовать в текущем или последующих проектах. 


In [40]:
#sbn.pairplot(df, hue = df['is_ultra'], kind = 'scatter')  
    


## Разбейте данные на выборки

In [50]:
df_train, df_valid = train_test_split(df, test_size=0.5, random_state=12345)
df_valid, df_test = train_test_split(df_valid, test_size=0.5, random_state=12345)
display(df_train.info())
display(df_valid.info())
df_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1607 entries, 2731 to 482
Data columns (total 5 columns):
calls       1607 non-null float64
minutes     1607 non-null float64
messages    1607 non-null float64
mb_used     1607 non-null float64
is_ultra    1607 non-null int64
dtypes: float64(4), int64(1)
memory usage: 75.3 KB


None

<class 'pandas.core.frame.DataFrame'>
Int64Index: 803 entries, 274 to 3197
Data columns (total 5 columns):
calls       803 non-null float64
minutes     803 non-null float64
messages    803 non-null float64
mb_used     803 non-null float64
is_ultra    803 non-null int64
dtypes: float64(4), int64(1)
memory usage: 37.6 KB


None

<class 'pandas.core.frame.DataFrame'>
Int64Index: 804 entries, 2499 to 677
Data columns (total 5 columns):
calls       804 non-null float64
minutes     804 non-null float64
messages    804 non-null float64
mb_used     804 non-null float64
is_ultra    804 non-null int64
dtypes: float64(4), int64(1)
memory usage: 37.7 KB


<div class="alert alert-block alert-success">
    
<b>Успех:</b> С разбиением все ок.
  </div>  
<div class="alert alert-block alert-info">
    
<b>Совет:</b> Не забывай проверять размеры полученных выборки,  это иной раз может предостеречь тебя от глупых ошибок.
  </div>  


<div class="alert alert-block alert-info">
    
<b>Комментарий студента:</b> Не нашел способа лучше чем вызвать инфо. С размером выборок все хорошо. Может быть есть сбособ лучше?
  </div>  

<div class="alert alert-block alert-success">
<b>Ответ[2]:</b> Нет тут достаточно info, или shape
</div>

## Исследуйте модели

In [42]:
#создаю переменные для признаков и целевого признака
features_train = df_train.drop(['is_ultra'], axis=1)
target_train = df_train['is_ultra']

features_valid = df_valid.drop(['is_ultra'], axis=1)
target_valid = df_valid['is_ultra']

features_test = df_test.drop(['is_ultra'], axis=1)
target_test = df_test['is_ultra']

### Модель Decision Tree (Дерево решений)

In [43]:
# Цикл для выбора лучшего значения параметра max_depth
for depth in range(4, 11,):
    model = DecisionTreeClassifier(random_state=12345, max_depth=depth) 
    model.fit(features_train, target_train) 
    predictions_valid = model.predict(features_valid) 
    print("max_depth =", depth, ": ", end='')
    print(accuracy_score(target_valid, predictions_valid))

max_depth = 4 : 0.7820672478206725
max_depth = 5 : 0.7820672478206725
max_depth = 6 : 0.788293897882939
max_depth = 7 : 0.788293897882939
max_depth = 8 : 0.7858032378580324
max_depth = 9 : 0.7795765877957659
max_depth = 10 : 0.7808219178082192


сперва сделал проверку на глубине дерева от 1 до 30 с шагом 3. По результатам лучший результат получился в диапазоне от 4 до 10. Перепроверил с новыми параметрами. Лучший результат вышел при max_depth = 6

### Модель Random Forest (Случайный лес)

In [44]:
# Цикл для выбора лучшего значения параметра n_estimators
for est in range(80, 101):
    model = RandomForestClassifier(random_state=12345, n_estimators= est, max_depth=6) 
    model.fit(features_train, target_train) 
    predictions_valid = model.predict(features_valid)
    print("n_estimators =", est, ": ", end='')
    print(accuracy_score(target_valid, predictions_valid))


n_estimators = 80 : 0.7920298879202988
n_estimators = 81 : 0.7907845579078456
n_estimators = 82 : 0.7920298879202988
n_estimators = 83 : 0.7907845579078456
n_estimators = 84 : 0.7920298879202988
n_estimators = 85 : 0.7907845579078456
n_estimators = 86 : 0.7932752179327521
n_estimators = 87 : 0.7920298879202988
n_estimators = 88 : 0.7920298879202988
n_estimators = 89 : 0.7907845579078456
n_estimators = 90 : 0.7907845579078456
n_estimators = 91 : 0.7920298879202988
n_estimators = 92 : 0.7932752179327521
n_estimators = 93 : 0.7920298879202988
n_estimators = 94 : 0.7932752179327521
n_estimators = 95 : 0.7920298879202988
n_estimators = 96 : 0.7932752179327521
n_estimators = 97 : 0.7907845579078456
n_estimators = 98 : 0.7907845579078456
n_estimators = 99 : 0.7907845579078456
n_estimators = 100 : 0.7895392278953923


Сперва прогнал выборку с количеством деревьев от 1 до 300 с шагом 10 и сделал вывод, что оптимально прогнать выборку от 80 до 100 с шагом 1. Так же  для достижения наилучшего результата установил max_depth=6 из предыдущего иследования. Лучший результат получился при n_estimators= 86

### Модель Logistic Regression (Логическая регрессия)

In [45]:
# Цикл для выбора лучшего значения параметра solver
solvers = ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga']
for i in solvers:
    model = LogisticRegression(solver= i, random_state=12345)
    model.fit(features_train, target_train) 
    predictions_valid = model.predict(features_valid)
    print("solvers =", i, ": ", end='')
    print(accuracy_score(target_valid, predictions_valid))

solvers = newton-cg : 0.7471980074719801
solvers = lbfgs : 0.7471980074719801
solvers = liblinear : 0.7484433374844334
solvers = sag : 0.7110834371108343
solvers = saga : 0.709838107098381


/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/sag.py:337: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
/opt/conda/lib/python3.7/site-packages/sklearn/linear_model/sag.py:337: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)


Лучший результат показан при solvers = liblinear. Не смог разобраться что за предупреждение выдает.


<div class="alert alert-block alert-info">

<b>Совет:</b> При каком-то из solver-ов,  смог алгоритм оптимизации не сошелся, это не критично, но говорит о том, что можно результат может быть посредственный, можно попробовать решить путем масштабирования или увеличив параметр `x_iter`
</div>

<div class="alert alert-block alert-info">
    
<b>Комментарий студента:</b> Что значит путем маштабирования? Нужно уменьшить выборки?
  </div>  

<div class="alert alert-block alert-success">
    
    
<b>Ответ[2]:</b> Изучите в дальнейших спринтах, можешь сам почитать пока об этом https://www.machinelearningmastery.ru/scale-standardize-or-normalize-with-scikit-learn-6ccc7d176a02/
</div> 

### Вывод
Модели со следущими параметрами дали лучший результат среди других параметров:

Decision Tree при max_depth = 6 accuracy = 0.788293897882939
Random Forest при depth = 6 и n_estimators = 86 accuracy = 0.7932752179327521
Logistic Regression при solver= liblinear accuracy = 0.7484433374844334

<div class="alert alert-block alert-success">

<b>Успех:</b> Хорошо проделан шаг. Радует корректное использование валидационной выборки 
</div>
<div class="alert alert-block alert-info">

<b>Совет:</b> Обрати также внимание на GridSearchCv и RandomizedSearchCV, эти инструменты могут быть удобными, когда большое количество параметров, уменьшает количество необходимых выборок, и делает код более лаконичным.
</div>

<div class="alert alert-block alert-info">
    
<b>Комментарий студента:</b> Спасибо за наводку. Прикольный метод только очень долгий. Возьму на вооружение.
  </div>  

<div class="alert alert-block alert-success">
    
    
<b>Ответ[2]:</b> Randomized -  быстрее
    
</div> 

In [54]:
# нахожу наилучшие параметры методом GridSearchCV
clf = RandomForestClassifier()
parametrs = { 'n_estimators': range (80, 100), 'max_depth': range (1,10)}

grid = GridSearchCV(clf, parametrs, cv=5)
grid.fit(features_train, target_train)
grid.best_params_

{'max_depth': 8, 'n_estimators': 96}

<div class="alert alert-block alert-info">
    
<b>Комментарий студента:</b> А почему результаты отличаются от моих выводов?
  </div>  

<div class="alert alert-block alert-success">
    
    
<b>Ответ[2]:</b> Потому что здесь трейн внутри разбивается на трейн_вадидацию 5 раз, за счет чего метрики другие и оценка более стабильная и объективная
</div> 

## Проверьте модель на тестовой выборке

In [47]:
# Для проверки тестовой выборки применю модель лучше всего себя проявившую в иследовании
model = RandomForestClassifier(max_depth=6, random_state=12345, n_estimators= 86)
model.fit(features_train, target_train)
predictions_test = model.predict(features_test)
accuracy_score(target_test, predictions_test)

0.8171641791044776

Проверка на тестовой выборке показала результат выше заданного проектом значения(0,75)

<div class="alert alert-block alert-success">

<b>Успех:</b> С тестированием тоже все в порядке, ты получил хорошее качество
</div>
<div class="alert alert-block alert-info">

<b>Совет:</b> Перед тестированием можно объединять валидационную и тренировочную выборку в одну обучающую, это может дать нам немного качества

<div class="alert alert-block alert-info">
    
<b>Комментарий студента:</b> Не совсем понял для чего это нужно делать. Ведь в начале специально датасет был поделен.
  </div>  

<div class="alert alert-block alert-success">
    
    
<b>Ответ[2]:</b> Он был поделен, чтобы настроить модели на валидационной выборке. А для итоговой оценки на тестовой выборке мы можем использовать остальные данные (тренировочную и валдиационную выборку)
</div> 

## (бонус) Проверьте модели на адекватность

In [48]:
model = DummyClassifier(strategy="most_frequent", random_state = 0) 
model.fit(features_train, target_train) 
predictions_valid = model.predict(features_valid)
print(accuracy_score(target_valid, predictions_valid))

0.709838107098381


<div class="alert alert-block alert-success">

<b>Успех:</b> Задание выполнено корректно, лайк за DummyClassifier - была выбрана правильная стратегия
</div>


<div class="alert alert-block alert-danger">

<b>Ошибка:</b> Не забывай про итоговый вывод - это важная важная часть любого проекта и заказчик обратит сюда особое внимание, здесь важно описать, что было сделано, рассмотренно, использовано, какаие результаты были достигнуты и т.п. <br> Вывод также важно структурировать и выделять ключевые. моменты

<div class="alert alert-block alert-info">
    
<b>Комментарий студента:</b> Торопился и упустил этот момент. Все исправлено вывод написан
  </div>  

## Общий вывод

1. Для работы были предоставленны данные оператора сотовой связи о поведении клиентов.
2. Данный представленны в виде таблицы состоящей и 5 столбцов и 3214 строк.
3. Предобработка данных не проводилась т.к. были предоставленны "чистые данные".
4. Для проведения исследования данные были разбиты на 3 выборки. Обучающую, валидную и тестовую в соотношении 50% 25% 25%.
5. Целевым признаком выбран столбец is_ultra.
6. Иследование было проведено со следующими гиперпараметрами:
       1. Модель Decision Tree (Дерево решений). сперва сделал проверку на глубине дерева от 1 до 30 с шагом 3. По результатам лучший результат получился в диапазоне от 4 до 10. Перепроверил с новыми параметрами. Лучший результат вышел при max_depth = 6
       2. Модель Random Forest (Случайный лес). Сперва прогнал выборку с количеством деревьев от 1 до 300 с шагом 10 и сделал вывод, что оптимально прогнать выборку от 80 до 100 с шагом 1. Так же для достижения наилучшего результата установил max_depth=6 из предыдущего иследования. Лучший результат получился при n_estimators= 86
       3. Модель Logistic Regression (Логическая регрессия). Лучший результат показан при solvers = liblinear.
7. Модели со следущими параметрами дали лучший результат среди других параметров:
       1. Decision Tree при max_depth = 6 accuracy = 0.788293897882939 
       2. Random Forest при depth = 6 и n_estimators = 86 accuracy = 0.7932752179327521 
       3. Logistic Regression при solver= liblinear accuracy = 0.7484433374844334
Для обучения лучше использовать модель Random Forest с гиперпараметрами depth = 6 и n_estimators = 86 accuracy = 0.7932752179327521
8. Проверка модели Random Forest, с лучшими гиперпараметрами, на тестовой выборке показала accuracy = 0.8171641791044776. Это хороший результат.
9. Проверка моделью  DummyClassifier дала accuracy = 0.709838107098381, а Random Forest accuracy = 0.7932752179327521 таким образом модель можно считать пригодной для использования.


<div class="alert alert-block alert-success">
    
    
<b>Успех[2]:</b> Всегда приятно наблюдать подробнный и структурированный итоговый вывод в конце работы!
</div> 

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x] Jupyter Notebook открыт
- [ ] Весь код исполняется без ошибок
- [ ] Ячейки с кодом расположены в порядке исполнения
- [ ] Выполнено задание 1: данные загружены и изучены
- [ ] Выполнено задание 2: данные разбиты на три выборки
- [ ] Выполнено задание 3: проведено исследование моделей
    - [ ] Рассмотрено больше одной модели
    - [ ] Рассмотрено хотя бы 3 значения гипепараметров для какой-нибудь модели
    - [ ] Написаны выводы по результатам исследования
- [ ] Выполнено задание 3: Проведено тестирование
- [ ] Удалось достичь accuracy не меньше 0.75
